Python/Product/PythonTools/PythonTools/Environments/EnvironmentSwitcherManager.cs

```diff
@@ -0,0 +1,277 @@
+                    _context.EnvironmentsChanged -= OnInterpretersChanged;
+                    _context.Dispose();
+                }
+
+                _context = value;
+
+                if (_context != null) {
+                    _context.EnvironmentsChanged += OnInterpretersChanged;
+                }
+            }
+        }
+
+        public void Initialize() {
+            if (_isInitialized) {
+                return;
+            }
+
+            _isInitialized = true;
+
+            _optionsService.DefaultInterpreterChanged += OnDefaultInterpreterChanged;
+            _registryService.InterpretersChanged += OnInterpretersChanged;
+            _workspaceService.OnActiveWorkspaceChanged += OnActiveWorkspaceChanged;
+            _monitorSelection.AdviseSelectionEvents(this, out _selectionEventsCookie);
+
+            SetInitialContext();
+        }
+
+        public void Dispose() {
+            if (_isInitialized) {
+                _optionsService.DefaultInterpreterChanged -= OnDefaultInterpreterChanged;
+                _registryService.InterpretersChanged -= OnInterpretersChanged;
+                _workspaceService.OnActiveWorkspaceChanged -= OnActiveWorkspaceChanged;
+                _monitorSelection.UnadviseSelectionEvents(_selectionEventsCookie);
+            }
+
+            if (Context != null) {
+                Context.EnvironmentsChanged -= OnInterpretersChanged;
+                Context.Dispose();
+            }
+        }
+
+        public string GetSwitcherCommandKeyBinding() {
+            // Don't call this during close because it throws an NRE
+            return IsClosing
+                ? null :
+                KeyBindingHelper.GetGlobalKeyBinding(GuidList.guidPythonToolsCmdSet, (int)PkgCmdIDList.cmdidViewEnvironmentStatus);
+        }
+
+        public Task SwitchToFactoryAsync(IPythonInterpreterFactory factory) {
+            return Context != null ? Context.ChangeFactoryAsync(factory) : Task.CompletedTask;
+        }
+
+        private Task OnActiveWorkspaceChanged(object sender, EventArgs e) {
+            _isPythonWorkspace = false;
+            RefreshFactories();
+            return Task.CompletedTask;
+        }
+
+        private void OnInterpretersChanged(object sender, EventArgs e) {
+            RefreshFactories();
+        }
+
+        private void OnDefaultInterpreterChanged(object sender, EventArgs e) {
+            RefreshFactories();
+        }
+
+        private void RefreshFactories() {
+            if (IsClosing) {
+                return;
+            }
+
+            AllFactories = Context?.AllFactories ?? Enumerable.Empty<IPythonInterpreterFactory>();
+            CurrentFactory = Context?.CurrentFactory;
+            EnvironmentsChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+        private void Reset() {
+            Context = null;
+            RefreshFactories();
+        }
+
+        private void SetInitialContext() {
+            IntPtr hierPtr = IntPtr.Zero;
```

Comment:
It is better to have a separate managed wrapper that works with `IntPtr` and use this wrapper in `EnvironmentSwitcherManager`.

Reply:
The wrapper class sounds overkill for this, but you made me look at the RTVS code at https://github.com/Microsoft/RTVS/blob/46a8f4579e4fd56219af3bb2cf3fd15b84464874/src/Package/Impl/ProjectSystem/ProjectSystemServices.cs#L47 and I realized that I can't let C# handle the container ptr out parameter that I don't care about, because it won't get freed.